grouper/fe/handlers.py

```diff
@@ -858,6 +886,188 @@ def get(self, group_id=None, name=None):
+                args_by_perm_json=json.dumps(args_by_perm))
+
+    def post(self, group_id=None, name=None):
+        group = Group.get(self.session, group_id, name)
+        if not group:
+            return self.notfound()
+
+        # only owner of group can request permissions for that group
+        role_index = self.current_user.my_role_index(group.my_members())
+        if role_index not in OWNER_ROLE_INDICES:
+            return self.forbidden()
+
+        # check inputs
+        form = GroupPermissionRequestForm(self.request.arguments)
+        args_by_perm = get_grantable_permissions(self.session)
+
+        form = self._prep_form(form, args_by_perm)
+
+        if not form.validate():
+            form = self._prep_form_for_display(form)
+            return self.render(
+                    "group-permission-request.html", form=form, group=group,
+                    args_by_perm_json=json.dumps(args_by_perm),
+                    alerts=self.get_form_alerts(form.errors),
+                    )
+
+        # load requested permission
+        if form.argument_select.data:
+            argument = form.argument_select.data
+        else:
+            argument = form.argument_text.data
+        permission = Permission.get(self.session, form.permission_name.data)
+
+        # make sure group doesn't already have this permission
+        for _, permission_name, _, permission_argument, _ in group.my_permissions():
+            if permission.name == permission_name and argument == permission_argument:
+                return self.render(
+                        "group-permission-request.html", form=form, group=group,
+                        args_by_perm_json=json.dumps(args_by_perm),
+                        alerts=[
+                            Alert("danger", "this group already has this permission+argument pair"),
+                            ],
+                        )
+
+        # make sure there's an owner for this
+        calc_args = args_by_perm.get(permission.name, [])
+        if not permission or (argument not in calc_args and "*" not in calc_args):
+            return self.render(
+                    "group-permission-request.html", form=form, group=group,
+                    args_by_perm_json=json.dumps(args_by_perm),
+                    alerts=[
+                        Alert("danger", "no such permission or permission+argument pair"),
+                        ],
+                    )
+
+        # save off request
+        try:
+            permissions.create_request(self.session, self.current_user, group,
+                    permission, argument, form.reason.data)
+        except permissions.RequestAlreadyExists:
+            alerts = [Alert("danger", "request for permission + argument already exists")]
+        except permissions.NoOwnersAvailable:
+            alerts = [Alert("danger", "no owners available for requested permission + argument")]
```

Comment:
If we're checking again at `create_request` time if the group already has the permission or if there are no owners available, does it make sense to do the explicit checks above?  I assume you're catching these exceptions because you could still be racing another caller, in which case maybe it would make sense to just try to create the request and catch the exceptions as the primary flow rather than pre-checking?  It just feels like the code is doing the same thing twice.